lib/core/job_cleanup.rb

```diff
@@ -0,0 +1,65 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'core'
+require 'utility/logger'
+
+module Core
+  class JobCleanUp
+    class << self
+      def execute(connector_id = nil)
+        cleanup_orphaned_jobs
+        cleanup_stuck_jobs(connector_id)
+      end
+
+      private
+
+      def cleanup_orphaned_jobs
+        Utility::Logger.info('Start cleaning up orphaned jobs...')
+        orphaned_jobs = ConnectorJob.orphaned_jobs
+        if orphaned_jobs.empty?
+          Utility::Logger.info('No orphaned jobs found. Skipping...')
+          return
+        end
+
+        # delete content indicies in case they are re-created by sync job
+        content_indices = orphaned_jobs.map(&:index_name).compact.uniq
+        ElasticConnectorActions.delete_indices(content_indices) if content_indices.any?
+        result = ConnectorJob.cleanup_jobs(orphaned_jobs)
+        Utility::Logger.error("Error found when deleting jobs: #{result['failures']}") if result['failures']&.any?
+        Utility::Logger.info("Successfully deleted #{result['deleted']} out of #{result['total']} orphaned jobs.")
+      end
+
+      def cleanup_stuck_jobs(connector_id = nil)
+        Utility::Logger.info("Start cleaning up stuck jobs for #{connector_id ? "connector #{connector_id}" : 'native connectors'}...")
+        stuck_jobs = ConnectorJob.stuck_jobs(connector_id)
+        if stuck_jobs.empty?
+          Utility::Logger.info('No stuck jobs found. Skipping...')
+          return
+        end
+
+        marked_count = 0
+        stuck_jobs.each do |job|
+          job.error!('The job has not seen any update for some time.')
+          Utility::Logger.debug("Successfully marked job #{job.id} as error.")
+
+          job_id = job.id
+          job = ConnectorJob.fetch_by_id(job_id)
+          Utility::Logger.warn("Could not found job by id #{job_id}") if job.nil?
```

Comment:
I wonder in what cases `job.id` is not presented.

Reply:
When the job is somehow deleted after `job.error!`.

`ConnectorJob.fetch_by_id` can return `nil`, so I want to be more defensive.

Reply:
I still don't get it. If I look at `ConnectorJob.stuck_jobs` implementation I see that you query the sync jobs index. This makes me think that all job entities have to have `id` because all Elastic documents have `id`.

Reply:
https://github.com/elastic/connectors-ruby/blob/61fd7d3a53e94ee995dadebb77aea49c09185dd6/lib/core/job_cleanup.rb#L48-L53

I'm not worrying about `job_id` in line 51 being `nil`. Line 53 is for the rare case `ConnectorJob.fetch_by_id(job_id)` returns `nil`, i.e. after `job.error!` in line 48, somehow the job doc is deleted. Because `fetch_by_id` can return `nil`, I want to be more defensive. In practice, this probably will never happen.

Reply:
Okay, here is my concern: if the code handles issues based on `In practice, this probably will never happen.` statement then we put ourselves in a situation where we just can't see the issues. From my point of view if we expect a specific behavior from our system then we trust it and if `neven happen` **does happen** then we should be able to see exceptions and we should fix them.

So I don't get why the job document can disappear after `job.error!` and if it really does, I'd like to fix that because it shouldn't :)